// redis/dist/lib/commands/HSET.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReply = exports.transformArguments = exports.FIRST_KEY_INDEX = void 0;
const generic_transformers_1 = require("./generic-transformers");
exports.FIRST_KEY_INDEX = 1;
function transformArguments(key, value) {
    const args = ['HSET', key];
    if (value instanceof Map) {
        pushMap(args, value);
    }
    else if (Array.isArray(value)) {
        pushTuples(args, value);
    }
    else if (typeof value === 'object' && value !== null) {
        pushObject(args, value);
    }
    return args;
}
exports.transformArguments = transformArguments;
function pushMap(args, map) {
    for (const [key, value] of map.entries()) {
        args.push(key.toString(), value.toString());
    }
}
function pushTuples(args, tuples) {
    args.push(...tuples.flat());
}
function pushObject(args, object) {
    for (const key of Object.keys(object)) {
        args.push(key.toString(), object[key].toString());
    }
}
exports.transformReply = generic_transformers_1.transformReplyString;
